### ISSUE TITLE
cmd/go: list -m -versions omits mismatched major versions

### ISSUE THREAD

**Issue by fransoaardi** [golang/go#25656]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
go version go1.10.2 darwin/amd64

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?
* replaced home directory as ${home}

GOARCH="amd64"
GOBIN=""
GOCACHE="${home} /Library/Caches/go-build"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="${home}/go"
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/x6/pxwr89_95yxbhw5nh8b7jk1w0000gn/T/go-build318943644=/tmp/go-build -gno-record-gcc-switches -fno-common"

### What did you do?
![image](https://user-images.githubusercontent.com/34496756/40758976-eac05694-64ca-11e8-9c74-5453f4a49285.png)

just typed 
```bash
vgo list -t github.com/go-chi/chi 
```
to get available releases but got only these two versions. 
Actually, go-chi/chi has latest release of v3.3.2

![image](https://user-images.githubusercontent.com/34496756/40759048-44b662ba-64cb-11e8-9ff5-2aa585fa9091.png)

If possible, provide a recipe for reproducing the error.
A complete runnable program is good.
A link on play.golang.org is best.

### What did you expect to see?
getting whole lists of available release

### What did you see instead?
just a part of available release

**Comment by myitcv** [comment]
I'll reference an answer I gave in https://github.com/golang/go/issues/25674#issuecomment-393782081

There are a couple of things at play here:

1. at the time of writing, `github.com/go-chi/chi` has not been converted to be a Go module
2. it has a major version `>= 2`

As a result therefore, what you are attempting won't, to my understanding, work until point 1 has been addressed.

In the meantime you will need to explicitly `vgo get github.com/go-chi/chi@v3.3.2` (or similar) (see linked comment)

**Comment by fransoaardi** [comment]
@myitcv 
Thanks for ur answer. 

In short, according to ur answer, the solution would be
```
1) to get the 3.3.2 version,  I need to explicitly write the version I want. 
2) go-chi/chi pkg generates go.mod
```

I wish I can get whole versions without go.mod pkg someday.  
Thanks.

**Comment by myitcv** [comment]
Here's a fully working example that uses v3.3.2.

Start by getting `vgo` in the usual way:

```
$ go get -u golang.org/x/vgo
```

Create a module using the `github.com/go-chi/chi` [example](https://github.com/go-chi/chi/tree/cca4135d8dddff765463feaf1118047a9e506b4a#examples):


```
$ mkdir hello
$ cd hello
$ cat <<EOD >hello.go
package main // import "example.com/hello"

import (
        "net/http"
        "github.com/go-chi/chi"
)

func main() {
        r := chi.NewRouter()
        r.Get("/", func(w http.ResponseWriter, r *http.Request) {
                w.Write([]byte("welcome"))
        })
        http.ListenAndServe(":3000", r)
}
EOD
```

Mark this as a module by creating a `go.mod`:


```
$ echo >go.mod
```

Now because, at the time of writing, `github.com/go-chi/chi`:

* has a major version `>= 2`
* has not been converted to a Go (vgo) module
* we want to use v3.3.2

we need to `vgo get` that specific version, which will be retrieved as a v0.0.0 psuedo version:


```
$ vgo get github.com/go-chi/chi@v3.3.2
vgo: finding github.com/go-chi/chi v0.0.0-20171222161133-e83ac2304db3
vgo: downloading github.com/go-chi/chi v0.0.0-20171222161133-e83ac2304db3
```

Now do a build to check all is good:


```
$ vgo build
```

And check the contents of `go.mod`:

```
$ cat go.mod
module example.com/hello

require github.com/go-chi/chi v0.0.0-20171222161133-e83ac2304db3
```

Version details for reference:

```
go version go1.10.2 linux/amd64 vgo:2018-02-20.1
vgo commit: 7c0cf33f39ccf5d8e4c59d34a150425ddac7df6a
```

**Comment by oiooj** [comment]
@fransoaardi You can run: `vgo list -t github.com/go-chi/chi/v3`

```
github.com/go-chi/chi/v3
	v3.0.0
	v3.1.0
	v3.1.1
	v3.1.2
	v3.1.3
	v3.1.4
	v3.1.5
	v3.2.0
	v3.2.1
	v3.3.0
	v3.3.1
	v3.3.2

```

**Comment by myitcv** [comment]
@oiooj you're quite right, thank you!

**Comment by myitcv** [comment]
@oiooj @fransoaardi - for reference, I've also added https://github.com/golang/go/issues/25712. As part of the description for that issue, I've asked whether we should perhaps have an issue that more broadly covers the requirements/etc for `vgo list`. One of the items in that list is giving `vgo list -t` (or some variant of it) the ability to pull back all versions across all major version numbers.

**Comment by rsc** [comment]
This is basically working as intended but unfortunate for older code that didn't know about the semantic import versioning conventions. Will need to think about whether to surface more here.

**Comment by bcl** [comment]
I think this is related: If a project didn't tag a release using a full semver version it won't be listed:

    vgo list -t github.com/gin-gonic/gin/
        v1.1.1
        v1.1.2
        v1.1.3
        v1.1.4

And if you try to get v1.2 it rewrites it as v1.2.0 and fails:

    vgo get github.com/gin-gonic/gin@v1.2                                                                                                      
    vgo get github.com/gin-gonic/gin: unknown revision "v1.2.0"

This may be working as designed, but I find it somewhat unexpected. I've asked gin them to add a v1.2.0 tag, but maybe vgo should handle it more gracefully?

**Comment by thinkerou** [comment]
The tag of gin usually is `v1.x.x`, but it have one release tag `v1.2` now, so I think vgo also should handle it, like tag also `hugo` and `gogs`.

**Comment by myitcv** [comment]
Just to point out that since https://go-review.googlesource.com/120198 `vgo list -t` is no longer an option. But I suspect that's a temporary state of affairs because this issue remains open and is the placeholder for some `vgo list` points (per https://github.com/golang/go/issues/25712#issuecomment-395211961)

**Comment by rsc** [comment]
The new form is list -m -versions and it still omits the mismatched major versions.

**Comment by rsc** [comment]
Mismatched major versions that are not using semantic import versioning (like in the example here) will be listed after the fix for #26238 lands, so closing this as a dup of #26238.

It's out of scope to list versions for other modules (foo/v2's versions when this is foo).
